sec-keys/openpgp-keys-bitcoin-core/openpgp-keys-bitcoin-core-20231014.ebuild

```diff
@@ -0,0 +1,20 @@
+# Copyright 1999-2023 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=8
+
+MyPN="guix.sigs"
```

Comment:
isn't this actually some bizarre list of people who just claim to have built it & got the same hash, rather than a signature from the release manager?

I remember looking into this about a year ago and being baffled as to how hard it was to find a proper keyring for the developers and also find a signature from them for the release tarball.

AFAICT (and I want to be wrong), anyone can get their key added there, and it only takes 1 valid signature for a pass.

Reply:
According to https://github.com/bitcoin/bitcoin/blob/ceb01689351e738436393cf7b8d84cb7fafc7b98/contrib/verify-binaries/README.md#preparation, we have to choose who to trust from the list.

Reply:
> we have to choose who to trust from the list.

Isn't this _always_ the case with verifying signatures? 🤔

Anyway, why does it matter if we add "extra" keys? The way PGP works, you can have keys for lots of random people in your keyring, but only the keys that were actually used to produce a signature will be used when verifying that signature. So doesn't it boil down to: trust that bitcoincore.org won't release a `SHA256SUMS.asc` file containing _only_ signatures by nefarious signers? Isn't that always the trust model we live with anyway?

Reply:
The difference is normally the list is a bunch of signers who are all core committers and we can just pick all-or-some of them, as opposed to having to filter out people who have nothing to do with Bitcoin development and just "attested" the release. I usually know the names of people doing releases for projects, as opposed to just collecting from a huge grabbag.

And yes, sort of, but it'd be nice to have a filtered list anyway, given the bar to attack would be "attest a release once, and compromise the website", which isn't that high of a bar considering -- espec. when we're comparing to the proper baseline which is whatever protection we get from someone just compromising the website.

Reply:
Okay, but then how does `verify-sig` _ever_ work for _any_ project? Are you saying that Gentoo packagers personally meet up with upstream project release managers to verify both their real-world identities and their PGP keys? Anything short of that, and we're just blindly trusting that the key we found in some repo somewhere is actually exclusively controlled by the same individual who develops the software being packaged.

For instance, let's say I know the name Wladimir van der Laan is prominently associated with Bitcoin Core. How do I, as a Gentoo packager, know what his PGP actually is? I can find a file named `laanwj.gpg`, but who knows where that really came from?

Reply:
Yes, that's the general principle of TOFU. You can at least verify that the same key is used to sign commits and regularly make releases going back years.

Allowing a release to be verified by someone who can just do a PR to add themselves in and it'll be accepted as standard practice is clearly a different situation.

This is clearly not the same thing as what most projects do, where you have a single key (or a handful) used for releases without third parties being encouraged to add their own.

Reply:
So is there a path forward, or do we abandon the idea of verifying sigs for Bitcoin Core? I could swear that _someone_ mentioned to me that they'd like to see Gentoo verifying the sigs, but maybe it wasn't you.

Reply:
> For instance, let's say I know the name Wladimir van der Laan is prominently associated with Bitcoin Core. How do I, as a Gentoo packager, know what his PGP key actually is?

That is a fundamental question to any cryptographic trust scheme. If I meet a person who says is laanwj, I don't now if the person really is laanwj. Even if the person is presenting some identification, I personally couldn't tell if it is genuine or forged.


> appear to have no substantive connection to the project to begin with

That is IMHO the main issue of this PR.


> Okay, but then how does `verify-sig` _ever_ work for _any_ project?

Some projects publish and use OpenPGP release keys. For example, for ejabberd (and other related projects) the vendor, Process One, publishes the key at https://www.process-one.net/downloads/KEYS, which we then use https://www.process-one.net/downloads/KEYS

This is what I would expect one reasonable precondition for verify-sig.


> So is there a path forward

Unless we find some official statement with release keys, I don't think there is a path forward at the moment.

Reply:
> For example, for ejabberd (and other related projects) the vendor, Process One, publishes the key at https://www.process-one.net/downloads/KEYS

But then making a malicious release requires only compromising the web server where both the release tarball and the `KEYS` file are hosted, so that is not "raising the bar" at all. If the compromise scenario we're trying to prevent is a website hack deploying a malicious tarball, then the same hack would also deploy a malicious `KEYS`. At least with the way Bitcoin Core is doing it, hosting the keys on GitHub, there's an extra step that an attacker must complete beyond just compromising bitcoincore.org.

Reply:
> But then making a malicious release requires only compromising the web server where both the release tarball and the `KEYS` file are hosted, so that is not "raising the bar" at all.

I would hope that some key rollover ceremony would take place in that case. How exactly such a ceremony would and should look like is another question.

Maybe we can get some sort of official statement from bitcoin-core devs that the release will always be signed by certain keys (fanquake certainly comes to mind)?

Reply:
> isn't this actually some bizarre list of people who just claim to have built it & got the same hash, rather than a signature from the release manager?

FWIW, this seems to be not accurate, as the keys include the ones of (presumably)
- laanwj https://github.com/bitcoin-core/guix.sigs/blob/main/builder-keys/laanwj.gpg
- fanquake https://github.com/bitcoin-core/guix.sigs/blob/main/builder-keys/fanquake.gpg
those keys also signed https://bitcoincore.org/bin/bitcoin-core-25.1/SHA256SUMS in https://bitcoincore.org/bin/bitcoin-core-25.1/SHA256SUMS.asc

fanquake did the latest release: https://github.com/bitcoin/bitcoin/releases/tag/v25.1

I am still torn on this, but if I am the only one who is skeptical, then we could give it a try…

Reply:
> I wonder if it wouldn't be better to just include fanquake's and laanwj's key and invoke gemato with `--no-require-all-good`

Concept ACK. At least we're reasonably certain that those two keys are really under the exclusive control of ostensibly honorable individuals. Requiring at least one sig from either of those two keys would serve adequately as a "sniff test" when bumping the ebuild in the future, which is how `verify-sig` is presented.

Maybe there ought to be a way to say to `verify-sig.eclass`: “Verify all the sigs that you can, but even if every sig in the `.asc` file verifies as good, still don't return success unless you got a good result from at least one of the following keys…”

Reply:
> But then making a malicious release requires only compromising the web server where both the release tarball and the `KEYS` file are hosted, so that is not "raising the bar" at all. If the compromise scenario we're trying to prevent is a website hack deploying a malicious tarball, then the same hack would also deploy a malicious `KEYS`. At least with the way Bitcoin Core is doing it, hosting the keys on GitHub, there's an extra step that an attacker must complete beyond just compromising bitcoincore.org.

Making a malicious release of projects where I personally am willing to add verify-sig support, would typically require compromising *all* of:
- the upstream software's main website
- the lead developer's personal blog
- github's associated-pgp-keys-for-a-given-user


And additionally requires me to look at the most recent update to sec-keys/openpgp-keys-*, wonder ***why on earth*** the PGP fingerprint has changed, but not follow up on that by getting in touch with the upstream maintainer, asking what happened, and almost certainly expecting that the new key is signed with a transition statement from the old key.

If a maintainer's website is randomly hacked and the publicly posted ascii-armored export of the canonical release signing key is replaced with a totally different key, and I'm the maintainer of the sec-keys package for that key, I'm very much not inclined to just "believe the web server". That is the entire point of having a signature verification process in the first place -- because I do not trust the server.

Reply:
Considering that bitcoin-core's attestation approach does not actually have anything to do with attesting that the sources are the official sources, but rather that the binaries are the same as someone else's binaries (???) I don't see that there's much value in verifying signatures even if signed by one of the bitcoin-core developers. What exactly are they attesting anyway? What, precisely, are bitcoin-core developers doing to validate that the thing they built is the source code that they developed and wrote and trust to be non-malicious?

Reply:
> > isn't this actually some bizarre list of people who just claim to have built it & got the same hash, rather than a signature from the release manager?
> 
> FWIW, this seems to be not accurate, as the keys include the ones of (presumably)
> 

I meant that the list includes others and regularly adds more. Not that it has nobody trustworthy therein. I thought that was clear from what I'd said but maybe not.

>     * laanwj [bitcoin-core/guix.sigs@`main`/builder-keys/laanwj.gpg](https://github.com/bitcoin-core/guix.sigs/blob/main/builder-keys/laanwj.gpg?rgh-link-date=2023-10-28T15%3A53%3A50Z)
> 
>     * fanquake [bitcoin-core/guix.sigs@`main`/builder-keys/fanquake.gpg](https://github.com/bitcoin-core/guix.sigs/blob/main/builder-keys/fanquake.gpg?rgh-link-date=2023-10-28T15%3A53%3A50Z)
>       those keys also signed [bitcoincore.org/bin/bitcoin-core-25.1/SHA256SUMS](https://bitcoincore.org/bin/bitcoin-core-25.1/SHA256SUMS) in [bitcoincore.org/bin/bitcoin-core-25.1/SHA256SUMS.asc](https://bitcoincore.org/bin/bitcoin-core-25.1/SHA256SUMS.asc)
> [...]
> …although I am still unsure about having such a large list if signatories. It appears that verify-sig requires that all signatories are known. I wonder if it wouldn't be better to just include fanquake's and laanwj's key and invoke gemato with `--no-require-all-good` (something the verify-sig.eclass does not support (yet)).

Yes, that would be different from using the repository as a keyring, and I think I could support that conceptually. Provided that someone actually attests the source code is what they uploaded, rather than just building binaries and if it matches someone else, signing it. It is not clear if that's actually the case.

Reply:
Maybe Bitcoin Core is an exceptional case since it deliberately *has no* "lead developer."

One thing is clear to me: if we're not going to (because we have no reliable way to) verify the source tarball downloaded from bitcoincore.org, then I would **strongly** prefer to fetch the source tarball from GitHub so we can be highly confident that it actually matches code that was tagged by someone with commit privileges on the repository. Downloading some sketch tarball from bitcoincore.org puts a lot of faith in that web host. (Someone working at the hosting company of that website is in a very good position to steal a lot of money.)

Reply:
That's an interesting point and I think we should consider that given the bits above wrt what the priorities _appear_ to be for the signatures on the website. Let me sleep on it a bit as well and see what others say.

Reply:
> Maybe Bitcoin Core is an exceptional case since it deliberately *has no* "lead developer."

It doesn't necessarily have to be a lead developer, for example Python has a defined process for appointing a release manager, and that decision is written up in a PEP and displayed on various sites and blogged about for over a year before the actual release of .0 (and then is responsible for regular releases for the next five years). That person *does* have the responsibility to PGP-sign the release tarballs for Python.

But I agree that Bitcoin Core's release process is definitely nonstandard, whatever the specifics.

> then I would strongly prefer to fetch the source tarball from GitHub so we can be highly confident that it actually matches code that was tagged by someone with commit privileges on the repository.

This is... actually a very good point and I think I strongly agree.

Reply:
> if we're not going to (because we have no reliable way to) verify the source tarball downloaded from bitcoincore.org, then I would **strongly** prefer to fetch the source tarball from GitHub so we can be highly confident that it actually matches code that was tagged by someone with commit privileges on the repository.

This seems to assume that something downloaded from github.org/<project>/releases is more trustworthy some something downloaded from a webpage. I doubt that this is true, with the information we have available. The webpage could be hosted on dedicated hardware in a secured space with only a handful of people having access to it. Whereas we do not know the number of people who could forge or modify a github release artifact. Presumable, at least some employees of github are able to do so. The argument that someone working at that company is in a very good position to steal a lot of money works here too.

I am also not sure if there are [signatures for github release artifacts of bitcoin](https://github.com/bitcoin/bitcoin/archive/refs/tags/v25.1.tar.gz). I could not find any.

Looking at the whole picture, I slightly favor the release artifact from bitcoincore.org. But ultimately, we obtain the most trust from ::gentoo's TOFU approach.

Reply:
> This seems to assume that something downloaded from github.org/<project>/releases is more trustworthy some something downloaded from a webpage. I doubt that this is true, with the information we have available. The webpage could be hosted on dedicated hardware in a secured space with only a handful of people having access to it. Whereas we do not know the number of people who could forge or modify a github release artifact. Presumable, at least some employees of github are able to do so. The argument that someone working at that company is in a very good position to steal a lot of money works here too.

But the flip side is that github employees have a vast, vast, vast wealth of projects that they could corrupt in order to steal money -- and I guarantee you, github is aware of this and has policies to prevent random employees from being able to directly access production data like that.

The value of using those links is that they are trustworthy in the sense that they directly correspond to -- and are automatically generated from -- the repository itself, using API endpoints that internally run `git archive ${ref}`. There is some guarantee that they will always directly represent the contents of the git repository from that tag, and forging this would require deploying a change to the github server codebase that special-cased a specific repository and postprocessed git-archive output. Note as well that github does NOT cache these tarballs, except temporarily -- so unless the theoretical malicious employee is either replacing it constantly by directly modifying caches (that humans should not have access to) or hiding production code in the server's VCS sources, you'd very shortly end up with a very auditable checksum mismatch.


> I am also not sure if there are [signatures for github release artifacts of bitcoin](https://github.com/bitcoin/bitcoin/archive/refs/tags/v25.1.tar.gz). I could not find any.

There are not. But the counterpoint is that the concern is, effectively, there are no signatures of the bitcoincore.org releases either.

Given a choice between two sources that effectively have no signatures, using the source that is easier to bit-for-bit reproduce by running git archive on the underlying git tags, is IMO more reliable.

Reply:
Hey, Bitcoin Core contributor and regular signatory of releases here. I'd like to shed some lights on some interrogations in this thread. I am not familiar with Gentoo's package management though.

First of all as @whitslack points out [since Wladimir stepped down](https://laanwj.github.io/2021/01/21/decentralize.html) there is no more lead maintainer of Bitcoin Core. There is nobody "in charge", processes are determined by a loosely defined group of regular contributors and [maintainers](https://github.com/bitcoin/bitcoin/blob/master/contrib/verify-commits/trusted-keys) enforce them.

Now it seems [there is confusion](https://github.com/gentoo/gentoo/pull/33555#discussion_r1375374394) between attestation of the sources corresponding to a release and attestation of a release's binaries.
To answer @eli-schwartz's question: the binaries signatures attest that the released binaries correspond to the released source compiled using a [bootstrappable toolchain](https://github.com/bitcoin/bitcoin/tree/master/contrib/guix).
Regarding the source, all commits on the master branch are [signed by one of the maintainers' key](https://github.com/bitcoin/bitcoin/tree/master/contrib/verify-commits). The closest you could get to an attestation of the sources for a release are:
- The signature of the merge commit tagged for this release ([here](https://github.com/bitcoin/bitcoin/commit/44d8b13c81e5276eb610c99f227a4d090cc532f6) for 26.0);
- The signature of the tag for this release ([26.0 tag](https://github.com/bitcoin/bitcoin/releases/tag/v26.0));
- The fact that all binaries attestations are based on this source. This is weaker as signatories do not guarantee they have reviewed the source, they merely guarantee the binaries correspond to the source. But presumably they wouldn't sign on to the release if the "official" source for this release was disputed.

> Downloading some sketch tarball from bitcoincore.org puts a lot of faith in that web host. (Someone working at the hosting company of that website is in a very good position to steal a lot of money.)

That's a tradeoff and as spelled-out above there are arguments for both sides. All i can tell is how the website is very carefully hosted and updated. Overall i would definitely trust more bitcoincore.org than the Github repository.

---

To conclude it seems the lowest hanging fruit to attesting the sources would be to verify the git tag signature if you can (usually, but not necessarily, made by fanquake these days).

Hope this helps!

Reply:
> First of all as @whitslack points out [since Wladimir stepped down](https://laanwj.github.io/2021/01/21/decentralize.html) there is no more lead maintainer of Bitcoin Core. There is nobody "in charge", processes are determined by a loosely defined group of regular contributors and [maintainers](https://github.com/bitcoin/bitcoin/blob/master/contrib/verify-commits/trusted-keys) enforce them.

This is, unfortunately, complete and utter garbage. I'm sorry, but it's true. This is one of several reasons why bitcoin, and the Bitcoin Core software, are a ludicrous joke to people outside your bubble.

You people are obsessed with "decentralization" as a marketing buzzword. You're so obsessed by it that you totally overlook the entire software development ecosystem, which already has a robust model for "decentralizing" the bus factor. Let me use the CPython programming language as an example here. They wanted to "reduce the bus factor" and "fairly represent all stakeholders" instead of terming it in some marketing buzzword. So, they:
- have a fairly large core developers team
- have a Steering Council regularly elected by the developers
- appoint different Release Managers on a regular basis to be in charge of creating releases, it's not bound to any one person



> Now it seems [there is confusion](https://github.com/gentoo/gentoo/pull/33555#discussion_r1375374394) between attestation of the sources corresponding to a release and attestation of a release's binaries.
> To answer @eli-schwartz's question: the binaries signatures attest that the released binaries correspond to the released source compiled using a [bootstrappable toolchain](https://github.com/bitcoin/bitcoin/tree/master/contrib/guix).
> Regarding the source, all commits on the master branch are [signed by one of the maintainers' key](https://github.com/bitcoin/bitcoin/tree/master/contrib/verify-commits). The closest you could get to an attestation of the sources for a release are:

No, I am not confused at all. I was asking a heavily sarcastic rhetorical question. Thanks for non-answering. :)

Since the closest we can get to an attestation of the sources for a release is to 

> * The signature of the merge commit tagged for this release ([here](https://github.com/bitcoin/bitcoin/commit/44d8b13c81e5276eb610c99f227a4d090cc532f6) for 26.0);
> 
> * The signature of the tag for this release ([26.0 tag](https://github.com/bitcoin/bitcoin/releases/tag/v26.0));

EITHER download the sources by downloading the ENTIRE git repository, then not verify the signature for the tag or merge commit since it turns out that Gentoo's verify-sig functionality doesn't implement verification of git commits, only distfiles, OR do no verification at all, which is already the status quo.

Option 3 is to have the ebuild maintainers manually, locally, check on their own machines that the PGP-signed git tag is valid and corresponds to the GitHub Inc. published `git-archive` tarballs, although without actually promising anything in the ebuild.

> * The fact that all binaries attestations are based on this source. This is weaker as signatories do not guarantee they have reviewed the source, they merely guarantee the binaries correspond to the source. But presumably they wouldn't sign on to the release if the "official" source for this release was disputed.

... but not this, since *once again* we don't care about this in the slightest. When you say "this is weaker" we say "this is so weak it's like water except all the water just evaporated".

It defeats the entire purpose of code signing. An ebuild already comes with a checksum manifest of the source tarball, PGP-signed by a Gentoo developer to authenticate that the Gentoo developer has not reviewed the source but is pretty sure it's probably the correct one. And of course, the Gentoo developer wouldn't sign on to the release if its validity was disputed.

> To conclude it seems the lowest hanging fruit to attesting the sources would be to verify the git tag signature if you can (usually, but not necessarily, made by fanquake these days).

This would definitely be effective, yes. It unfortunately has four problems:

- it is 300mb instead of a mere 12mb
- Gentoo would have to grow infrastructure for verifying a git tag signature as it currently doesn't exist
- Gentoo frowns on the use of git repositories as sources for stable releases instead of "live ebuilds" that always build the latest commit
- use of git for retrieving sources is inherently less secure than use of tarball distfiles, because tarball distfiles are checksummed using a secure blake2 hash whereas git is locked to sha1 which has had several well publicized weaknesses in recent years and is believed to be "hopefully still secure but people should really stop using it"

The lowest hanging fruit is actually for "usually, but not necessarily, fanquake" to

Reply:
I just stumbled upon this thread and figured i'd give a hand. You seem emotional about this, so i won't engage further. I'll just mention this:
1. Nobody mentioned "decentralization" nor any other buzzword here. It's simply a matter of fact: there is no authority attesting sources for a given version. If you do want one for your package manager, maybe step up and do it yourself instead of whining others won't do it for you?
2. As for the ludicrous comparison with CPython, i doubt in their case anyone who would appear to have any sort of influence on the development of the project would get [legally harassed](https://bitcoindefense.org/our-work/) by well funded attackers (and more).

Reply:
> Nobody mentioned "decentralization" nor any other buzzword here. It's simply a matter of fact: there is no authority attesting sources for a given version.

You mentioned it. You linked to https://laanwj.github.io/2021/01/21/decentralize.html

> One thing is clear: this is a serious project now, and we need to start taking decentralization seriously.

> Plans
> > Decentralize the release process, and release signing.
> > > Rename the GPG key used to sign SHA256SUMS.asc to “Bitcoin Core release signing key”, instead of having it in my personal title. Make some construct so that N of M (minimally) trusted gitian signers doing a succesful build automatically results in a signed distribution.

There is your buzzword, from your own citation by your former lead.

> If you do want one for your package manager, maybe step up and do it yourself instead of whining others won't do it for you?

And there's why it is impossible to take bitcoin seriously.

Why should anyone on planet Earth trust an arbitrary bystander to "do it yourself instead of whining others won't do it for you" in order to provide AUTHORSHIP metadata for the bitcoin software?

I'm not the author of a bitcoin release, so how am I supposed to PGP-sign it as the AUTHOR of a release?

There are innumerable things one can step up and do yourself in FOSS, but "claim to be the author of the software" is not a serious suggestion on your part.

> As for the ludicrous comparison with CPython, i doubt in their case anyone who would appear to have any sort of influence on the development of the project would get [legally harassed](https://bitcoindefense.org/our-work/) by well funded attackers (and more).

What does this have to do with the concept of a consensus governance model ("decentralized" for those who like buzzwords)? If thousands of developers of bitcoin are already being legally harassed, how does drawing up a documented model for thousands of people getting an equal vote make it more likely that one person will "have any sort of influence on the development of the project"?

If you already have someone who is "basically always the person who releases a new version", how does documenting that that person is the standard release manager make that person more legally harassed than being the person who is already associated with every release? (This isn't even influence on the development of the project, this isn't *development*, this is *release*.)

Again, this is not about changing how bitcoin is developed or released, it's about having the person that does the release *anyway*, sign that release as a testimonial "I am one of several dozen core developers and I have verified this tarball is the real deal".

That person is *already* signing the git tag, just not the tarball file. That's the entire problem from start to finish, from a consumer perspective.

Reply:
> I just stumbled upon this thread and figured i'd give a hand. You seem emotional about this, so i won't engage further. I'll just mention this:

(Also note: I didn't originally intend to post that comment, it was in draft mode but somehow github accidentally submitted it. I was considering to tone it down in a second pass, but then the damage was done so I wasn't sure how beneficial it would be to do so at that point. T)

Reply:
> Regarding the source, all commits on the master branch are [signed by one of the maintainers' key](https://github.com/bitcoin/bitcoin/tree/master/contrib/verify-commits).

Thanks for the pointer to this directory. It seems the best we could do at the moment is to use https://github.com/bitcoin/bitcoin/blob/master/contrib/verify-commits/trusted-keys as OpenPGP trust anchor for source-code release artifacts. Or am I missing something?

Reply:
@Flowdalic,

We cannot use that to verify archives, because there are no archives that are signed by the release manager.

What we could do is use git-r3 to download sources, and submit patches to verify-sig.eclass which allows it to verify the signatures on a git tag or commit object.

This would allow verifying (albeit at sha1 hash strength instead of blake2) the identity of the release.

The trusted-keys file is not new information that has been uncovered by this thread. It is simply not topical to the proposal in this PR.

Again, the initial objecting review by @thesamesam back in October was that there are no signatures on source code archives, and even trusted committers don't check the source code, they only download an existing release from GitHub (that may or may not have been compromised) and add a PGP signature testifying that the anonymous source code archive was compliant with https://reproducible-builds.org when built by different people but with the same guix VM snapshot.

It has nothing to do with authorship and the entire process very specifically disavows all claim of knowing that the code is the real bitcoin code.

They explicitly don't care if the Bitcoin code has a malicious backdoor as long as everyone compiles the same backdoor.

Reply:
> The trusted-keys file is not new information that has been uncovered by this thread.

This particular trusted-keys file was new to me. I do not remember it being discussed in this PR so far. But then again, the discussion here got pretty lengthy, so I may have missed it. In that case, I am sorry.

> We cannot use that to verify archives, because there are no archives that are signed by the release manager.

Are you sure about that?

https://bitcoincore.org/bin/bitcoin-core-26.0/SHA256SUMS.asc  contains signatures of two keys found in https://github.com/bitcoin/bitcoin/blob/60b6ff5ac030b65d63deda18517a6a5a02ead3b8/contrib/verify-commits/trusted-keys:
- D1DBF2C4B96F2DEBF4C16654410108112E7EA81F
- 152812300785C96444D3334D17565732E08E5E41

This `SHA256SUMS` file, which is authenticated by `SHA256SUMS.asc`, contains, among others, the SHA-256 of bitcoin-26.0.tar.gz.

@darosior Is there a guarantee from bitcoin-core, that at least one key in `contrib/verify-commits/trusted-keys` will be used to sign the SHA256SUMS published on https://bitcoincore.org/bin/?

I am not saying we should use this `trusted-keys` file as trust anchor, because it would somehow be a circular trust relation. Ideally, the trusted keys would be published in a versioned manner on bitcoincore.org. However, as far as I can tell, it is the best thing we have as trust anchor so far. But please correct me if I am wrong.

Reply:
> I am not saying we should use this `trusted-keys` file as trust anchor, because it would somehow be a circular trust relation. Ideally, the trusted keys would be published in a versioned manner on bitcoincore.org. However, as far as I can tell, it is the best thing we have as trust anchor so far. But please correct me if I am wrong.

There's nothing wrong with using a trust anchor from the source code it signs??? Why would there be.

It's just keydata, the trusted information itself is the fingerprints and anyways the point is always to manually check trust anchors and then assume they are correct.

It has nothing to do with the problem here.

> Are you sure about that?

Yes, I'm very sure of it.

Again: signing means nothing here.

Signing means that they blindly accepted someone else's tarball without checking what it is.

This is not trustworthy signing.

There is no solving this, not while using distfiles. The ONLY way to solve it is to use git-r3.eclass